### PR TITLE
Implement support for token's in NuGetVersion

### PR DESCRIPTION
**Pull request by clairernovotny**
## Bug
Fixes: https://github.com/NuGet/Home/issues/6228
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: Adds support for parsing and formatting token values in NuGetVersion. Also adds a validation check on package creation that version is not tokenized.

## Testing/Validation
Tests Added: Yes
Reason for not adding tests:  
Validation done:  Tests run

Possibly needs additional tests (need guidance for what to add, etc)

### CODE REVIEW

**Comment by clairernovotny** [comment]
@emgarten @unniravindranathan @rrelyea @joelverhagen @rohit21agrawal

**Comment by emgarten** [comment]
This would affect restore. If a package is created incorrectly and still has tokenized versions after packing it looks like they could end up being read, compared according to the new rules, and used in restore since NuGetVersion supports them, but the result would be nonsensical.

**Comment by clairernovotny** [comment]
@emgarten I can add validations there to block a restore as an error, would that work?

**Comment by clairernovotny** [comment]
Alternatively to adding extra checks, how about adding another overload to Parse/TryParse on VersionRange and NuGetVersion that defaults to false that controls this? 

It’d be similar to how allowFloating is there:
        public static bool TryParse(string value, bool allowFloating, out VersionRange versionRange)

I could add 
        public static bool TryParse(string value, bool allowFloating, bool allowToken, out VersionRange versionRange)

and have the defaults as false. Then the Manifest.ReadFrom path can supply a true/false for allowing this behavior that’d percolate through?

That would localize the changes to callers that have explicitly asked for it?

**Comment by rohit21agrawal** [comment]
that sounds like a much better (and safer ) approach to me.

**Comment by clairernovotny** [comment]
@rohit21agrawal @emgarten I've added overloads to control the use of tokens in the code-paths that read the nuspec values. The defaults are all false, so no existing code should be impacted by this.

File: src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs [hunk 1]
```diff
@@ -205,6 +213,11 @@ public XDocument Xml
         }
 
         public virtual PackageIdentity GetIdentity()
+        {
+            return GetIdentity(false);
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs]: use allowToken: false for readability

File: src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs [hunk 2]
```diff
@@ -87,9 +87,17 @@ public virtual string GetId()
         /// Version of the package
         /// </summary>
         public virtual NuGetVersion GetVersion()
+        {
+            return GetVersion(false);
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs]: allowToken: false

File: src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs [hunk 3]
```diff
@@ -205,6 +213,11 @@ public XDocument Xml
         }
 
         public virtual PackageIdentity GetIdentity()
+        {
+            return GetIdentity(false);
+        }
+
+        public virtual PackageIdentity GetIdentity(bool allowToken)
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs]: where is allowToken being used in this function?

File: src/NuGet.Core/NuGet.Packaging/NuGet.Packaging.csproj [hunk 4]
```diff
@@ -49,6 +49,11 @@
   </ItemGroup>
   
   <ItemGroup>
+    <Compile Update="PackageCreation\Resources\NuGetResources.Designer.cs">
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging/NuGet.Packaging.csproj:52]: what are these changes for?

File: src/NuGet.Core/NuGet.Packaging/NuspecReader.cs [hunk 5]
```diff
@@ -113,6 +113,14 @@ public IEnumerable<PackageDependencyGroup> GetDependencyGroups()
...
         /// </summary>
         public IEnumerable<PackageDependencyGroup> GetDependencyGroups(bool useStrictVersionCheck)
+        {
+            return GetDependencyGroups(useStrictVersionCheck: useStrictVersionCheck, allowTokens: false);
+        }
+
+        /// <summary>
+        /// Read package dependencies for all frameworks
+        /// </summary>
+        public IEnumerable<PackageDependencyGroup> GetDependencyGroups(bool useStrictVersionCheck, bool allowTokens)
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging/NuspecReader.cs:123]: default this to false so as to minimize changes in callers

Comment (clairernovotny) [reply src/NuGet.Core/NuGet.Packaging/NuspecReader.cs:123]: It's already defaulted by way of the other overload, does that work?

File: src/NuGet.Core/NuGet.Packaging/NuGet.Packaging.csproj (hunk 4 above)
Comment (clairernovotny) [reply src/NuGet.Core/NuGet.Packaging/NuGet.Packaging.csproj:52]: That was redone by VS when setting the ResX to regenerate the strongly typed files. I added a resource for an exception

File: src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs (hunk 3 above)
Comment (clairernovotny) [reply src/NuGet.Core/NuGet.Packaging.Core/NuspecCoreReaderBase.cs]: Oops, should have been in GetVersion, will fix.

**Comment by rrelyea** [comment]
Would prefer to see a preprocessing XmlReader that replaces the tokens with something else (or something similar), rather than making our key data classes enabled for designer scenarios.

**Comment by clairernovotny** [comment]
@rrelyea there's no amount of pre-processing that will allow these to round-trip without essentially duplicating the entire object model... ?

And Tokens are in the nuspec spec, just not currently supported by the readers, hence this PR. I'd think there should be a way to model them in the data classes since the data classes today cannot fully represent the spec.

**Comment by rrelyea** [comment]
do you want to preserve an xml comment in the .nuspec?

generally, as I found from my time on WPF/xaml working with blend and cider ... design time and run time have very different needs. I don't want to try to morph nuget types to fully support design time.

**Comment by clairernovotny** [comment]
Agreed that there are always limitations and designers have different needs. That said, shouldn't the core object model fully be able to represent a valid nuspec?

**Comment by rohit21agrawal** [comment]
@onovotny i would suggest instead of modifying the NuGetVersion data model , you modify the ManifestMetadata class to take in a RawVersionString .

You can then modify the call @ https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/ManifestReader.cs#L80 to set the RawVersionString in the ManifestMetadata if NuGetVersion.TryParse returns a false value.

Even this should also be only done with overloads that specify whether you are reading a tokenized nuspec or a well formed nuspec.

**Comment by clairernovotny** [comment]
@rohit21agrawal that's not the only place tokens are used -- they can be used on a package dependency, within a version range.... we have to be able to parse and display that correctly too.

**Comment by rohit21agrawal** [comment]
NuGetVersion class should only be able to create an object if it's a  NuGetVersion as the name suggests. We cannot break that.

**Comment by clairernovotny** [comment]
@rohit21agrawal I very much disagree. It should support the Nuspec data model, which includes tokens. It may not be a valid version, but that's why there's a property called `IsTokenized`. Versions and the token representation are a core component. Right now it cannot be represented correctly in the object model.

It may not be a valid "runtime version," but it is a valid Nuspec version outside of a package.

**Comment by clairernovotny** [comment]
As an alternative, it could all be subclassed as `TokenizedNuGetVersion`, which is read from the Manifest/Nuspec reader paths. The effect would be very similar though since all of the places that use the new NuGetVersion.Parse overload would need to return the alternate type instead.

I think Subclassing it hides more since it's not immediately obvious from looking at the API that it might be a tokenized instance unless you do a type check on it.

**Comment by rrelyea** [comment]
Should the WPF runtime XAML parser be able to roundtrip all markup extensions.
We said no.

I think this is a similar situation

**Comment by emgarten** [comment]
`NuGetVersion` is a concrete, resolved version. Other things such as `1.0.*` are also not allowed in `NuGetVersion` and it would be difficult if the contract for `NuGetVersion` changed everywhere to need an extra check to make sure that the version passed in wasn't a tokenized placeholder in places where something else parsed the version and passed it in. It feels equivalent to allowing `null` everywhere for versions.

For the NPE scenario I would expect `ManifestReader` to only use strings to allow these invalid values that are replaced later. Validation could be done as a separate optional step. Would that work for this @onovotny @rohit21agrawal?

**Comment by clairernovotny** [comment]
If `ManifestReader` only allowed strings for version, along with PackageDependency, that could work. We'd still need `Manifest.Save` to work as well with a string in those places.

We'd still need to copy (which I hate) VersionRange, and related types, since we'd need to properly display and format the package dependency data.

**Comment by clairernovotny** [comment]
Looking more into using strings for ManifestReader, that appears to be a much larger, change as it means we need to also update IPackageMetadata and PackageDependency to support a custom VersionRange string that then needs to be specially parsed (and round-tripped correctly).

I'm not sure this will really work.

I draw a distinction between `1.0.*` and `$version$`. One is a range and one is the value of a version. `1.*` is allowed anywhere `VersionRange` is and is controlled by the `allowFloating` parameter on its parse methods.

`$version$` is no different for `NuGetVersion` than `*` is for VersionRange. Allowed in some contexts and disallowed by others. It's literally the representation of the data model. Just because it hasn't been needed by the current code paths doesn't mean that it's invalid. It just means you haven't had to fully represent the nuspec file in your object model yet.

The WPF/XAML analogies don't really hold here either. This isn't arbitrary user code, it's a token value that can be explicitly used in Nuspec fields and is supported by some operations today (pack) as replaceable items. That usage allowed for a shortcut in being able to replace them before reading into the OM. That just side-stepped the issue of actually representing data (since there wasn't a need to yet).

**Comment by clairernovotny** [comment]
Worked around by https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/pull/264

**Pull request closed without merging**